include/exadg/incompressible_navier_stokes/time_integration/time_int_bdf_coupled_solver.cpp

```diff
@@ -241,6 +241,28 @@ TimeIntBDFCoupled<dim, Number>::do_timestep_solve()
     if(this->param.right_hand_side)
       pde_operator->evaluate_add_body_force_term(rhs, this->get_next_time());
 
+    // Add the convective term to the right-hand side of the equations
+    // if the convective term is treated explicitly (additive decomposition):
+    // evaluate convective term and add extrapolation of convective term to the rhs (-> minus sign!)
+    if(this->param.convective_problem() and
+       this->param.treatment_of_convective_term == TreatmentOfConvectiveTerm::Explicit)
+    {
+      if(this->param.ale_formulation)
+      {
+        // evaluate convective term for all previous times since the mesh has been updated
+        for(unsigned int i = 0; i < this->vec_convective_term.size(); ++i)
+        {
+          // in a general setting, we only know the boundary conditions at time t_{n+1}
+          pde_operator->evaluate_convective_term(this->vec_convective_term[i],
+                                                 solution[i].block(0),
+                                                 this->get_next_time());
+        }
+      }
```

Comment:
Where does this code come from? This should only be implemented once. Lines 245-260 (with another closing bracket) can be done before the `if(nonlinear)-else` statement in this function.

Only the two lines below, adding the convective terms to the `rhs` vector may be written redundantly for the linear and nonlinear case.

I would suggest to do this change in a separate PR, since it is a pure refactoring of the existing code, but might potentially break the current code version (if we introduce a mistake here).

Reply:
I pulled out the update of the mesh velocity at previous timesteps and the computation of `sum_alpha_ui` as they are both identical for the two code paths.